src/librustc_expand/expand.rs

```diff
@@ -363,7 +363,17 @@ impl<'a, 'b> MacroExpander<'a, 'b> {
                 krate.attrs = vec![];
                 krate.module = ast::Mod { inner: orig_mod_span, items: vec![], inline: true };
             }
-            _ => unreachable!(),
+            Some(ast::Item { span, kind, .. }) => {
+                krate.attrs = vec![];
+                krate.module = ast::Mod { inner: orig_mod_span, items: vec![], inline: true };
+                self.cx.span_err(
+                    span,
+                    &format!(
+                        "expected crate top-level item to be a module after macro expansion, found a {}",
```

Comment:
Crate-level attribute macros aren't really expected to produce a module, they are just not supported at all and may result in nonsense and paradoxes.
This is an ok fix and it removes the ICE, but this case should ideally report the same error as this example:
```rust
enum E {
    #[test] // error: expected an inert attribute, found an attribute macro
    V
}

fn main() {}
```

Reply:
It's not urgent though, and I still hope to rewrite the macro invocation collector anyway.